File: src/test/java/com/github/jactor/web/interceptor/RequestInterceptorTest.java
package com.github.jactor.web.interceptor;

import static com.github.jactor.web.interceptor.RequestInterceptor.CHOSEN_LANGUAGE;
import static com.github.jactor.web.interceptor.RequestInterceptor.CURRENT_REQUEST;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.Mockito.when;

import com.github.jactor.web.Language;
import com.github.jactor.web.LanguageKt;
import com.github.jactor.web.Request;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import javax.servlet.http.HttpServletRequest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.i18n.LocaleContextHolder;
import org.springframework.web.servlet.ModelAndView;

@SpringBootTest
@DisplayName("A RequestInterceptor")
class RequestInterceptorTest {

  @Autowired
  private RequestInterceptor requestInterceptorToTest;
  @MockBean
  private HttpServletRequest httpServletRequestMock;

  @BeforeEach
  void mockRequest() {
    when(httpServletRequestMock.getRequestURI()).thenReturn("/page");
    when(httpServletRequestMock.getQueryString()).thenReturn("some=param");
  }

  @Test
  @DisplayName("should add current url without language to the model")
  void shouldAddCurrentUrlWithoutLangugaeToTheModel() {
    ModelAndView modelAndView = new ModelAndView();

    when(httpServletRequestMock.getRequestURI()).thenReturn("/somewhere");
    when(httpServletRequestMock.getQueryString()).thenReturn("out=there&lang=something&another=param");

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Map<String, Object> model = modelAndView.getModel();

    var currentRequest = Optional.ofNullable(model.get(CURRENT_REQUEST));

    assertThat(currentRequest).hasValueSatisfying(request-> assertAll(
        () -> assertThat(request).isInstanceOf(Request.class),
        () -> assertThat((Request) request).extracting(Request::getCurrentUrl).isEqualTo("/somewhere?out=there&another=param")
    ));
  }

  @Test
  @DisplayName("should add Norwegian language to model")
  void shouldAddNorwegianLanguageToModel() {
    LocaleContextHolder.setLocale(new Locale("no"));
    ModelAndView modelAndView = new ModelAndView();

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Language language = (Language) modelAndView.getModel().getOrDefault(CHOSEN_LANGUAGE, new Language(new Locale("svada"), "there"));

    assertAll(
        () -> assertThat(language.getName()).as("name").isEqualTo(LanguageKt.NORSK),
        () -> assertThat(language.getLocale()).isEqualTo(new Locale("no"))
    );
  }

  @Test
  @DisplayName("should add English language to model")
  void shouldAddEnglishLanguageToModel() {
    LocaleContextHolder.setLocale(new Locale("en"));
    ModelAndView modelAndView = new ModelAndView();

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Language language = (Language) modelAndView.getModel().getOrDefault(CHOSEN_LANGUAGE, new Language(new Locale("svada"), "there"));

    assertAll(
        () -> assertThat(language.getName()).as("name").isEqualTo(LanguageKt.ENGLISH),
        () -> assertThat(language.getLocale()).isEqualTo(new Locale("en"))
    );
  }

  @Test
  @DisplayName("should add Thai language to model")
  void shouldAddThaiLanguageToModel() {
    LocaleContextHolder.setLocale(new Locale("th"));
    ModelAndView modelAndView = new ModelAndView();

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Language language = (Language) modelAndView.getModel().getOrDefault(CHOSEN_LANGUAGE, new Language(new Locale("svada"), "there"));

    assertAll(
        () -> assertThat(language.getName()).as("name").isEqualTo(LanguageKt.THAI),
        () -> assertThat(language.getLocale()).isEqualTo(new Locale("th"))
    );
  }

  @Test
  @DisplayName("should add English language to model when locale is not supported")
  void shouldAddEnglishLanguageToModelWhenLocaleIsUnsupported() {
    ModelAndView modelAndView = new ModelAndView();

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Language language = (Language) modelAndView.getModel().getOrDefault(CHOSEN_LANGUAGE, new Language(new Locale("svada"), "there"));

    assertAll(
        () -> assertThat(language.getName()).as("name").isEqualTo(LanguageKt.ENGLISH),
        () -> assertThat(language.getLocale()).isEqualTo(new Locale("en"))
    );
  }

  @Test
  @DisplayName("should add Thai language to model from language parameters")
  void shouldAddThaiLanguageToModelFromParameters() {
    LocaleContextHolder.setLocale(new Locale("no"));
    ModelAndView modelAndView = new ModelAndView();

    when(httpServletRequestMock.getQueryString()).thenReturn("select=something&lang=th");

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Language language = (Language) modelAndView.getModel().getOrDefault(CHOSEN_LANGUAGE, new Language(new Locale("svada"), "there"));

    assertAll(
        () -> assertThat(language.getName()).as("name").isEqualTo(LanguageKt.THAI),
        () -> assertThat(language.getLocale()).isEqualTo(new Locale("th"))
    );
  }

  @Test
  @DisplayName("should add English language to model from language parameters")
  void shouldAddEnglishLanguageToModelFromParameters() {
    LocaleContextHolder.setLocale(new Locale("no"));
    ModelAndView modelAndView = new ModelAndView();

    when(httpServletRequestMock.getQueryString()).thenReturn("select=something&lang=en");

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Language language = (Language) modelAndView.getModel().getOrDefault(CHOSEN_LANGUAGE, new Language(new Locale("svada"), "there"));

    assertAll(
        () -> assertThat(language.getName()).as("name").isEqualTo(LanguageKt.ENGLISH),
        () -> assertThat(language.getLocale()).isEqualTo(new Locale("en"))
    );
  }

  @Test
  @DisplayName("should add Norwegian language to model from language parameters")
  void shouldAddNorwegianLanguageToModelFromParameters() {
    LocaleContextHolder.setLocale(new Locale("en"));
    ModelAndView modelAndView = new ModelAndView();

    when(httpServletRequestMock.getQueryString()).thenReturn("select=something&lang=no");

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Language language = (Language) modelAndView.getModel().getOrDefault(CHOSEN_LANGUAGE, new Language(new Locale("svada"), "there"));

    assertAll(
        () -> assertThat(language.getName()).as("name").isEqualTo(LanguageKt.NORSK),
        () -> assertThat(language.getLocale()).isEqualTo(new Locale("no"))
    );
  }

  @Test
  @DisplayName("should add English language to model when unknown language parameter")
  void shouldAddEnglishLanguageToModelWhenUnknownLanguageParameter() {
    LocaleContextHolder.setLocale(new Locale("th"));
    ModelAndView modelAndView = new ModelAndView();

    when(httpServletRequestMock.getQueryString()).thenReturn("select=something&lang=fi");

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Language language = (Language) modelAndView.getModel().getOrDefault(CHOSEN_LANGUAGE, new Language(new Locale("svada"), "there"));

    assertAll(
        () -> assertThat(language.getName()).as("name").isEqualTo(LanguageKt.ENGLISH),
        () -> assertThat(language.getLocale()).isEqualTo(new Locale("en"))
    );
  }

  @Test
  @DisplayName("should add chosen view to the model")
  void shouldAddChosenViewToTheModel() {
    ModelAndView modelAndView = new ModelAndView();
    when(httpServletRequestMock.getRequestURI()).thenReturn("/user");

    requestInterceptorToTest.postHandle(httpServletRequestMock, null, null, modelAndView);

    Map<String, Object> model = modelAndView.getModel();

    var currentRequest = Optional.ofNullable(model.get(CURRENT_REQUEST));

    assertThat(currentRequest).hasValueSatisfying(request-> assertAll(
        () -> assertThat(request).isInstanceOf(Request.class),
        () -> assertThat((Request) request).extracting(Request::getChosenView).isEqualTo("user")
    ));
  }
}

File: src/test/kotlin/com/github/jactor/web/interceptor/RequestInterceptorTest.kt
package com.github.jactor.web.interceptor

import com.github.jactor.web.ENGLISH
import com.github.jactor.web.Language
import com.github.jactor.web.NORSK
import com.github.jactor.web.Request
import com.github.jactor.web.THAI
import com.github.jactor.web.interceptor.RequestInterceptor.Companion.CHOSEN_LANGUAGE
import com.github.jactor.web.interceptor.RequestInterceptor.Companion.CURRENT_REQUEST
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll
import org.mockito.kotlin.whenever
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.context.i18n.LocaleContextHolder
import org.springframework.web.servlet.ModelAndView
import java.util.Locale

@SpringBootTest
@DisplayName("A RequestInterceptor")
class RequestInterceptorTest {

    @Autowired
    private lateinit var requestInterceptorToTest: RequestInterceptor

    @MockBean
    private lateinit var httpServletRequestMock: HttpServletRequest

    @MockBean
    private lateinit var httpServletResponseMock: HttpServletResponse

    private val handler: Any = Object()

    @BeforeEach
    fun `mock request`() {
        whenever(httpServletRequestMock.requestURI).thenReturn("/page")
        whenever(httpServletRequestMock.queryString).thenReturn("some=param")
    }

    @Test
    fun `should add current url without language to the model`() {
        val modelAndView = ModelAndView()

        whenever(httpServletRequestMock.requestURI).thenReturn("/somewhere")
        whenever(httpServletRequestMock.queryString).thenReturn("out=there&lang=something&another=param")

        requestInterceptorToTest.postHandle(httpServletRequestMock, httpServletResponseMock, handler, modelAndView)

        val model = modelAndView.model
        val currentRequest = model[CURRENT_REQUEST]

        assertAll(
            { assertThat(currentRequest).isInstanceOf(Request::class.java) },
            { assertThat(currentRequest as Request).extracting(Request::currentUrl).isEqualTo("/somewhere?out=there&another=param") }
        )
    }

    @Test
    fun `should add Norwegian language to model`() {
        LocaleContextHolder.setLocale(Locale("no"))
        val modelAndView = ModelAndView()

        requestInterceptorToTest.postHandle(httpServletRequestMock, httpServletResponseMock, handler, modelAndView)

        val language = modelAndView.model.getOrDefault(CHOSEN_LANGUAGE, Language(Locale("svada"), "there")) as Language

        assertAll(
            { assertThat(language.name).`as`("name").isEqualTo(NORSK) },
            { assertThat(language.locale).isEqualTo(Locale("no")) }
        )
    }

    @Test
    @DisplayName("")
    fun `should add English language to model`() {
        LocaleContextHolder.setLocale(Locale("en"))
        val modelAndView = ModelAndView()

        requestInterceptorToTest.postHandle(httpServletRequestMock, httpServletResponseMock, handler, modelAndView)

        val language = modelAndView.model.getOrDefault(CHOSEN_LANGUAGE, Language(Locale("svada"), "there")) as Language

        assertAll(
            { assertThat(language.name).`as`("name").isEqualTo(ENGLISH) },
            { assertThat(language.locale).isEqualTo(Locale("en")) }
        )
    }

    @Test
    fun `should add Thai language to model`() {
        LocaleContextHolder.setLocale(Locale("th"))
        val modelAndView = ModelAndView()

        requestInterceptorToTest.postHandle(httpServletRequestMock, httpServletResponseMock, handler, modelAndView)

        val language = modelAndView.model.getOrDefault(CHOSEN_LANGUAGE, Language(Locale("svada"), "there")) as Language

        assertAll(
            { assertThat(language.name).`as`("name").isEqualTo(THAI) },
            { assertThat(language.locale).isEqualTo(Locale("th")) }
        )
    }

    @Test
    fun `should add Thai language to model from language parameters`() {
        LocaleContextHolder.setLocale(Locale("no"))
        val modelAndView = ModelAndView()
        whenever(httpServletRequestMock.queryString).thenReturn("select=something&lang=th")

        requestInterceptorToTest.postHandle(httpServletRequestMock, httpServletResponseMock, handler, modelAndView)

        val language = modelAndView.model.getOrDefault(CHOSEN_LANGUAGE, Language(Locale("svada"), "there")) as Language

        assertAll(
            { assertThat(language.name).`as`("name").isEqualTo(THAI) },
            { assertThat(language.locale).isEqualTo(Locale("th")) }
        )
    }

    @Test
    fun `should add English language to model from language parameters`() {
        LocaleContextHolder.setLocale(Locale("no"))
        val modelAndView = ModelAndView()
        whenever(httpServletRequestMock.queryString).thenReturn("select=something&lang=en")

        requestInterceptorToTest.postHandle(httpServletRequestMock, httpServletResponseMock, handler, modelAndView)

        val language = modelAndView.model.getOrDefault(CHOSEN_LANGUAGE, Language(Locale("svada"), "there")) as Language

        assertAll(
            { assertThat(language.name).`as`("name").isEqualTo(ENGLISH) },
            { assertThat(language.locale).isEqualTo(Locale("en")) }
        )
    }

    @Test
    fun `should add Norwegian language to model from language parameters`() {
        LocaleContextHolder.setLocale(Locale("en"))
        val modelAndView = ModelAndView()
        whenever(httpServletRequestMock.getQueryString()).thenReturn("select=something&lang=no")

        requestInterceptorToTest.postHandle(httpServletRequestMock, httpServletResponseMock, handler, modelAndView)

        val language = modelAndView.model.getOrDefault(CHOSEN_LANGUAGE, Language(Locale("svada"), "there")) as Language

        assertAll(
            { assertThat(language.name).`as`("name").isEqualTo(NORSK) },
            { assertThat(language.locale).isEqualTo(Locale("no")) }
        )
    }

    @Test
    fun `should add English language to model when unknown language parameter`() {
        LocaleContextHolder.setLocale(Locale("th"))
        val modelAndView = ModelAndView()
        whenever(httpServletRequestMock.getQueryString()).thenReturn("select=something&lang=fi")

        requestInterceptorToTest.postHandle(httpServletRequestMock, httpServletResponseMock, handler, modelAndView)

        val language = modelAndView.model.getOrDefault(CHOSEN_LANGUAGE, Language(Locale("svada"), "there")) as Language

        assertAll(
            { assertThat(language.name).`as`("name").isEqualTo(ENGLISH) },
            { assertThat(language.locale).isEqualTo(Locale("en")) }
        )
    }

    @Test
    fun `should add chosen view to the model`() {
        val modelAndView = ModelAndView()
        whenever(httpServletRequestMock.requestURI).thenReturn("/user")

        requestInterceptorToTest.postHandle(httpServletRequestMock, httpServletResponseMock, handler, modelAndView)

        val model = modelAndView.model
        val currentRequest = model[CURRENT_REQUEST]

        assertAll(
            { assertThat(currentRequest).isInstanceOf(Request::class.java) },
            { assertThat(currentRequest as Request).extracting(Request::chosenView).isEqualTo("user") }
        )
    }
}
